#------------------------------ tabstop = 4 ----------------------------------
#
# Copyright (C) 2024 Comcast
#
# All rights reserved.
#
# This software is protected by copyright laws of the United States
# and of foreign countries. This material may also be protected by
# patent laws of the United States and of foreign countries.
#
# This software is furnished under a license agreement and/or a
# nondisclosure agreement and may only be used or copied in accordance
# with the terms of those agreements.
#
# The mere transfer of this software does not imply any licenses of trade
# secrets, proprietary technology, copyrights, patents, trademarks, or
# any other form of intellectual property whatsoever.
#
# Comcast retains all ownership rights.
#
#------------------------------ tabstop = 4 ----------------------------------

#
# Created by Kevin Funderburg on 09/11/2024
#

From 07e7293ab10b6397557b9b6b76256a901d8fa90e Mon Sep 17 00:00:00 2001
From: Jesse Crews <jesse_crews@comcast.com>
Date: Wed, 20 Jan 2021 10:59:44 -0600
Subject: [PATCH 1/3] gnulib: Update frexp[l] from upstream

This change pulls frexp and frexpl modules from upstream
and when the system does not provide frexp, the gnulib implementation is
used.

 * Unset gl_cv_func_frexpl_broken_beyond_repair when frexp[l] are not
   present to allow substitution
 * Initialize gl_cv_func_frexp_decl to false on !have_frexpl
 * Initialize gl_cv_func_ldexpl_decl to false on !have_ldexpl
---
 glib/gnulib/frexp.c     | 172 +++++++++++++++++++++++++++++++++++++---
 glib/gnulib/frexpf.c    |  26 ++++++
 glib/gnulib/frexpl.c    |  48 +++++++----
 glib/gnulib/meson.build |   6 +-
 4 files changed, 220 insertions(+), 32 deletions(-)
 create mode 100644 glib/gnulib/frexpf.c

diff --git a/glib/gnulib/frexp.c b/glib/gnulib/frexp.c
index c26d82c88..288ca66fe 100644
--- a/glib/gnulib/frexp.c
+++ b/glib/gnulib/frexp.c
@@ -1,22 +1,168 @@
-#include <config.h>
+/* Split a double into fraction and mantissa.
+   Copyright (C) 2007-2021 Free Software Foundation, Inc.
+
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <https://www.gnu.org/licenses/>.  */
+
+/* Written by Paolo Bonzini <bonzini@gnu.org>, 2003, and
+   Bruno Haible <bruno@clisp.org>, 2007.  */
+
+#if ! defined USE_LONG_DOUBLE
+# include <config.h>
+#endif

 /* Specification.  */
-#include <gnulib_math.h>
+#include <math.h>

 #include <float.h>
-#include "isnand-nolibm.h"
+#ifdef USE_LONG_DOUBLE
+# include "isnanl-nolibm.h"
+# include "fpucw.h"
+#else
+# include "isnand-nolibm.h"
+#endif

-double rpl_frexp (double x, int *expptr)
+/* This file assumes FLT_RADIX = 2.  If FLT_RADIX is a power of 2 greater
+   than 2, or not even a power of 2, some rounding errors can occur, so that
+   then the returned mantissa is only guaranteed to be <= 1.0, not < 1.0.  */
+
+#ifdef USE_LONG_DOUBLE
+# define FUNC frexpl
+# define DOUBLE long double
+# define ISNAN isnanl
+# define DECL_ROUNDING DECL_LONG_DOUBLE_ROUNDING
+# define BEGIN_ROUNDING() BEGIN_LONG_DOUBLE_ROUNDING ()
+# define END_ROUNDING() END_LONG_DOUBLE_ROUNDING ()
+# define L_(literal) literal##L
+#else
+# define FUNC frexp
+# define DOUBLE double
+# define ISNAN isnand
+# define DECL_ROUNDING
+# define BEGIN_ROUNDING()
+# define END_ROUNDING()
+# define L_(literal) literal
+#endif
+
+DOUBLE
+FUNC (DOUBLE x, int *expptr)
 {
-  if (x == 0.0 || x == -0.0)
+  int sign;
+  int exponent;
+  DECL_ROUNDING
+
+  /* Test for NaN, infinity, and zero.  */
+  if (ISNAN (x) || x + x == x)
+    {
+      *expptr = 0;
+      return x;
+    }
+
+  sign = 0;
+  if (x < 0)
+    {
+      x = - x;
+      sign = -1;
+    }
+
+  BEGIN_ROUNDING ();
+
   {
-    *expptr = x;
-    return x;
+    /* Since the exponent is an 'int', it fits in 64 bits.  Therefore the
+       loops are executed no more than 64 times.  */
+    DOUBLE pow2[64]; /* pow2[i] = 2^2^i */
+    DOUBLE powh[64]; /* powh[i] = 2^-2^i */
+    int i;
+
+    exponent = 0;
+    if (x >= L_(1.0))
+      {
+        /* A positive exponent.  */
+        DOUBLE pow2_i; /* = pow2[i] */
+        DOUBLE powh_i; /* = powh[i] */
+
+        /* Invariants: pow2_i = 2^2^i, powh_i = 2^-2^i,
+           x * 2^exponent = argument, x >= 1.0.  */
+        for (i = 0, pow2_i = L_(2.0), powh_i = L_(0.5);
+             ;
+             i++, pow2_i = pow2_i * pow2_i, powh_i = powh_i * powh_i)
+          {
+            if (x >= pow2_i)
+              {
+                exponent += (1 << i);
+                x *= powh_i;
+              }
+            else
+              break;
+
+            pow2[i] = pow2_i;
+            powh[i] = powh_i;
+          }
+        /* Avoid making x too small, as it could become a denormalized
+           number and thus lose precision.  */
+        while (i > 0 && x < pow2[i - 1])
+          {
+            i--;
+            powh_i = powh[i];
+          }
+        exponent += (1 << i);
+        x *= powh_i;
+        /* Here 2^-2^i <= x < 1.0.  */
+      }
+    else
+      {
+        /* A negative or zero exponent.  */
+        DOUBLE pow2_i; /* = pow2[i] */
+        DOUBLE powh_i; /* = powh[i] */
+
+        /* Invariants: pow2_i = 2^2^i, powh_i = 2^-2^i,
+           x * 2^exponent = argument, x < 1.0.  */
+        for (i = 0, pow2_i = L_(2.0), powh_i = L_(0.5);
+             ;
+             i++, pow2_i = pow2_i * pow2_i, powh_i = powh_i * powh_i)
+          {
+            if (x < powh_i)
+              {
+                exponent -= (1 << i);
+                x *= pow2_i;
+              }
+            else
+              break;
+
+            pow2[i] = pow2_i;
+            powh[i] = powh_i;
+          }
+        /* Here 2^-2^i <= x < 1.0.  */
+      }
+
+    /* Invariants: x * 2^exponent = argument, and 2^-2^i <= x < 1.0.  */
+    while (i > 0)
+      {
+        i--;
+        if (x < powh[i])
+          {
+            exponent -= (1 << i);
+            x *= pow2[i];
+          }
+      }
+    /* Here 0.5 <= x < 1.0.  */
   }
-  else if (isnan (x))
-    return x;
-  else if (isinf (x))
-    return x;
-#undef frexp
-  return frexp (x, expptr);
+
+  if (sign < 0)
+    x = - x;
+
+  END_ROUNDING ();
+
+  *expptr = exponent;
+  return x;
 }
diff --git a/glib/gnulib/frexpf.c b/glib/gnulib/frexpf.c
new file mode 100644
index 000000000..8fec6b8fd
--- /dev/null
+++ b/glib/gnulib/frexpf.c
@@ -0,0 +1,26 @@
+/* Split a float into fraction and mantissa.
+   Copyright (C) 2011-2021 Free Software Foundation, Inc.
+
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <https://www.gnu.org/licenses/>.  */
+
+#include <config.h>
+
+/* Specification.  */
+#include <math.h>
+
+float
+frexpf (float x, int *expptr)
+{
+  return (float) frexp ((double) x, expptr);
+}
diff --git a/glib/gnulib/frexpl.c b/glib/gnulib/frexpl.c
index cd6a81c77..a5b1fd98e 100644
--- a/glib/gnulib/frexpl.c
+++ b/glib/gnulib/frexpl.c
@@ -1,21 +1,35 @@
+/* Split a 'long double' into fraction and mantissa.
+   Copyright (C) 2007, 2009-2021 Free Software Foundation, Inc.
+
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <https://www.gnu.org/licenses/>.  */
+
 #include <config.h>
-#include <gnulib_math.h>
-#include <float.h>
-#include <math.h>

-long double rpl_frexpl (long double x, int *expptr)
+#if HAVE_SAME_LONG_DOUBLE_AS_DOUBLE
+
+/* Specification.  */
+# include <math.h>
+
+long double
+frexpl (long double x, int *expptr)
 {
-  if (x == 0.0L || x == -0.0L)
-  {
-    *expptr = x;
-    return x;
-  }
-  else if (isnanl (x))
-    return x;
-  else if (isinf (x))
-    return x;
-#ifndef _MSC_VER
-#undef frexpl
-#endif
-  return frexpl (x, expptr);
+  return frexp (x, expptr);
 }
+
+#else
+
+# define USE_LONG_DOUBLE
+# include "frexp.c"
+
+#endif
diff --git a/glib/gnulib/meson.build b/glib/gnulib/meson.build
index 38b530aa0..b38d078b4 100644
--- a/glib/gnulib/meson.build
+++ b/glib/gnulib/meson.build
@@ -293,13 +293,14 @@ if have_frexp
   subdir ('gl_cv_func_frexp_works')
 else
   gl_cv_func_frexp_works = false
-  gl_cv_func_frexp_broken_beyond_repair = true
+  gl_cv_func_frexp_broken_beyond_repair = false
 endif
 if have_frexpl
   subdir ('gl_cv_func_frexpl_works')
 else
   gl_cv_func_frexpl_works = false
-  gl_cv_func_frexpl_broken_beyond_repair = true
+  gl_cv_func_frexpl_broken_beyond_repair = false
+  gl_cv_func_frexpl_decl = false
 endif

 if not gl_cv_func_frexp_works and gl_cv_func_frexp_broken_beyond_repair
@@ -321,6 +322,7 @@ if have_ldexpl
   subdir ('gl_cv_func_ldexpl_works')
 else
   gl_cv_func_ldexpl_works = false
+  gl_cv_func_ldexpl_decl = false
 endif
 math_h_config.set ('REPLACE_LDEXPL', gl_cv_func_ldexpl_works ? 0 : 1)
 math_h_config.set ('HAVE_DECL_LDEXPL', gl_cv_func_ldexpl_decl ? 0 : 1)
--
2.27.0

